Use PySNMP >6 asyncio API

Emantor opened the pull request:
**Description**
Update pysnmp and it's users to use the new async API, this should fix CI and also ensures we can properly use pysnmp in the future.
Pushed additional fixes for no longer required pylint disables.

**Checklist**
- [x] PR has been tested

Fixes https://github.com/labgrid-project/labgrid/issues/1493
Closes #1456

Emantor committed:
util: convert SimpleSNMP to asyncio

Wrap the now asyncio based co-routines in loop.run_until_complete().

Signed-off-by: Rouven Czerwinski <r.czerwinski@pengutronix.de>
Signed-off-by: Bastian Krause <bst@pengutronix.de>

Emantor committed:
pyproject: allow pysnmp newer than 6.x

With the new asyncio based handling we can unlock the pysnmp dependency.
The lexstudio fork has also taken over maintenance of pysnmp on pypi, so
switch back to pysnmp.

Signed-off-by: Rouven Czerwinski <r.czerwinski@pengutronix.de>

Emantor committed:
resouce: ethernetport: convert to asyncio

Follow the sync API deprecation and use ensure_event_loop() making sure
we have an asyncio loop even if no loop can be retrieved.

Signed-off-by: Rouven Czerwinski <r.czerwinski@pengutronix.de>
Signed-off-by: Bastian Krause <bst@pengutronix.de>

Bastian-Krause commented on labgrid/util/snmp.py:
```diff
@@ -8,24 +10,39 @@ def __init__(self, host, community, port=161):
+        try:
+            # if called from async code, try to get current's thread loop
+            self.loop = asyncio.get_running_loop()
+        except RuntimeError:
+            # no previous, external or running loop found, create a new one
+            self.loop = asyncio.new_event_loop()
```
Client code does the same, but only in `start_session()` when no loop is provided otherwise (and stashes the loop for later use).

How would other code retrieve the newly created event loop from here?

I think this must rely on an existing event loop, similar to EthernetPortManager:

https://github.com/labgrid-project/labgrid/blob/6b541210c6063e97d81e257d2d6bda1444d9ec78/labgrid/resource/ethernetport.py#L251

Emantor replied:
But `get_event_loop` is deprecated, so I chose not to use this.

Bastian-Krause replied:
No, `get_event_loop()` is not deprecated per se: [only if there is no current event loop](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.get_event_loop). That's why I think we should simply expect a loop here. If there is none, that's an error that needs to be fixed outside of labgrid.

Emantor replied:
Using labgrid as a library means that there is not necessarily an available loop and we need to support this (which is exactly the case where there is no loop).

Emantor replied:
Testing this inside of labgrid client reveals that `get_running_loop` does not find the running client event loop and instead no loop is found and another is created for SNMP.

Bastian-Krause replied:
> Using labgrid as a library means that there is not necessarily an available loop and we need to support this (which is exactly the case where there is no loop).

I disagree. There are certain "entry points" into labgrid where a loop can be passed or labgrid tries to find one or sets a new one up as a fallback. Exporter, coordinator and client are these "entry points". Client and libraries typically use `labgrid.remote.client.start_session()` to start a session.  So that makes sure there is a loop.

For all other cases, code that uses labgrid as a library needs to make sure a loop exists. Maybe that needs to be documented better.

Using `get_event_loop()` in EthernetPortManager was an active decision. Using that makes sure there is an event loop. Otherwise you get a warning (and a new event loop) or an error, depending on the Python version.

> Testing this inside of labgrid client reveals that `get_running_loop` does not find the running client event loop and instead no loop is found and another is created for SNMP.

That's because `get_running_loop()` does not work in sync contexts and there is no other way to achieve that.

Emantor replied:
I guess this is fine for the SNMP switch handling, but isn't the correct requirement for the simple power backend handling we had beforehand. So how do we proceed with this? Requiring an event loop for sync library usage is not really feasible.

Bastian-Krause replied:
How does the simple power backend handling work without a session?

Emantor replied:
> How does the simple power backend handling work without a session?

What do you mean? You can totally describe all of the `NetworkPowerPort`s inside of the YAML configuration and not use a `RemotePlace` to retrieve the resources.

Bastian-Krause replied:
Yes, I see.

See commit messages of da95c444, 0e1e4fba and 606af353 for context.

What about calling `labgrid.remote.client.ensure_event_loop()` whenever we need an event loop? That would restore the original behavior of using labgrid as a library without a coordinator session.

We could add a function that allows library users to pass their event loop into labgrid (stashing it) for use cases without `labgrid.remote.client.start_session()` (which already allows passing an external loop).

Bastian-Krause commented on labgrid/util/snmp.py:
```diff
@@ -8,24 +10,39 @@ def __init__(self, host, community, port=161):
+    def cleanup(self):
+        self.engine.closeDispatcher()
+        self.loop.run_until_complete(self.loop.shutdown_asyncgens())
+        self.loop.close()
```
Doesn't this close a potential external loop?

Emantor replied:
Ah, yes, we should probably stash the fact that we created a loop and only close this if we did.

Emantor replied:
Implemented.

Bastian-Krause committed:
remote/client: move ensure_event_loop() to labgrid.util.loop

Moving the function and the ContextVar to labgrid.util.loop will allow
non-client code to use it as well. Otherwise usage will result in
circular imports such as:

      from labgrid import Target
  labgrid/__init__.py:1: in <module>
      from .target import Target
  labgrid/target.py:10: in <module>
      from .driver import Driver
  labgrid/driver/__init__.py:4: in <module>
      from .serialdriver import SerialDriver
  labgrid/driver/serialdriver.py:10: in <module>
      from ..util.proxy import proxymanager
  labgrid/util/proxy.py:7: in <module>
      from ..resource.common import Resource
  labgrid/resource/__init__.py:2: in <module>
      from .ethernetport import SNMPEthernetPort
  labgrid/resource/ethernetport.py:9: in <module>
      from ..remote.client import ensure_event_loop
  labgrid/remote/client.py:41: in <module>
      from .. import Environment, Target, target_factory
  E   ImportError: cannot import name 'Environment' from partially initialized module 'labgrid' (most likely due to a circular import) (labgrid/__init__.py)

Signed-off-by: Bastian Krause <bst@pengutronix.de>

Bastian-Krause committed:
util/loop: add is_new_loop()

There is currently no way to find out whether ensure_event_loop() created
a new loop. This information is useful in scenarios where a resource is
created without a session in a sync context, because then it needs to
shut down the newly created loop in clean up code. But this must not
happen if a running loop is used.

Since ensure_event_loop() is already public API people use, add a
separate function checking whether the passed in loop is identical to
the loop stashed in the loop creation code in ensure_event_loop().

Signed-off-by: Bastian Krause <bst@pengutronix.de>

Bastian-Krause commented:
[Reworked](https://github.com/labgrid-project/labgrid/compare/871f4844e7d3b44a5d5a0c8559853e8df8990f82..9c7cb236b32a5298d85e9e278c7a79dee2420261):
- move `ensure_event_loop()` to `labgrid.util.loop` for usage outside client code
- add `is_new_loop()` to check whether `ensure_event_loop()` created a new loop
- use `ensure_event_loop()` and `is_new_loop()` in `labgrid.util.snmp` and `labgrid.resource.ethernetport`

Bastian-Krause commented:
@Emantor Did you have a look at the rework?

Bastian-Krause committed:
doc/conf: add pysnmp.hlapi.v3arch.asyncio to autodoc_mock_imports

Signed-off-by: Bastian Krause <bst@pengutronix.de>
